### PR TITLE
Virtual Member Functions and Exceptions

### PR DESCRIPTION
**Pull request by blakemulnix**
This pull requests adds virtual member functions and the tests and documentation for that feature. 

This pull request also adds exceptions and the tests for that feature. The documentation for exception still needs to be completed.

### CODE REVIEW

File: cmake/cmakepp_core/exceptions/exceptions.cmake [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    cpp_assert_signature("${ARGV}" desc)
+
+    # Get the map of exception handlers and attempt to get the handler for
+    # this type from the map
+    cpp_get_global(_r_exception_handlers "_CPP_EXCEPTION_HANDLERS_")
+    cpp_map(GET "${_r_exception_handlers}" _r_fxn_2_call "${_r_exec_type}")
+
+    # Call the function if its found, else throw an error
+    if("${_r_fxn_2_call}" STREQUAL "")
+        message(FATAL_ERROR "No exception handler for exception type ${_r_exec_type}")
```
Comment (ryanmrichard) [inline cmake/cmakepp_core/exceptions/exceptions.cmake:97]: You probably want to make the message something more like `Uncaught ${_r_exec_type} exception: ${ARGV}"`

**Comment by blakemulnix** [comment]
I added documentation for exceptions and I updated the documentation for maps to include all maps functions.

**Comment by ryanmrichard** [comment]
It just occurred to me that we need a way to register an exception handler which can handle any exception (the equivalent of C++'s `catch(...){ /* handler code */ };`

File: cmake/cmakepp_core/exceptions/exceptions.cmake [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            message(FATAL_ERROR "This function expects at least one argument")
+        endif()
+
+        foreach(_etc_arg_i ${ARGN})
+            cpp_assert_type("desc" "${_etc_arg_i}")
+        endforeach()
+    endif()
+
+    foreach(_etc_exce_type_i ${ARGN})
+        cpp_assert_signature("${ARGV}" desc)
```
Comment (ryanmrichard) [inline cmake/cmakepp_core/exceptions/exceptions.cmake:171]: left over?

File: docs/source/getting_started/cmakepp_examples/index.rst [hunk 3]
```diff
@@ -14,5 +14,6 @@ CMakePP automatically. If you have not done this step, please see
    :numbered:
 
    classes
+   exceptions
```
Comment (ryanmrichard) [inline docs/source/getting_started/cmakepp_examples/index.rst:17]: Is the warning from the `docs_test` action valid? (does `getting_started/cmakepp_examples/exceptions` exist?)

File: docs/source/getting_started/cmakepp_examples/maps.rst [hunk 4]
```diff
@@ -62,3 +62,53 @@ constructor in the following way:
...
+
+We can copy a map using the ``COPY`` keyword:
+
+.. code-block:: cmake
+
+  # Copy my_map to my_map_copy
+  cpp_map(COPY "${my_map}" my_map_copy)
+
+Checking Equality of Maps
+=========================
```
Comment (ryanmrichard) [inline docs/source/getting_started/cmakepp_examples/maps.rst:87]: It'd be nice to resolve the duplicate label warnings in this file.

**Pull request merged**